tsz_test.go

```diff
@@ -212,31 +148,70 @@ func TestConcurrentRoundtrip10MsBetweenWrites(t *testing.T) {
 	testConcurrentRoundtrip(t, 10*time.Millisecond)
 }
 
-// poorly synchronized. we want to test reading while writing, after finishing writing
-// we could introduce explicit synchronization but i don't see a simple way to do so
-// in a way that doesn't reduce the possible timing overlaps.
+// Test reading while writing at the same time.
 func testConcurrentRoundtrip(t *testing.T, sleep time.Duration) {
-
 	s := New(TwoHoursData[0].t)
-	done := make(chan struct{})
-	errors := make(chan string)
 
-	wg := &sync.WaitGroup{}
+	//notify the reader about the number of points that have been written.
+	writeNotify := make(chan int)
+
+	// notify the reader when we have finished.
+	done := make(chan struct{})
 
-	go func() {
-		for err := range errors {
-			t.Error(err)
+	// continuously iterate over the values of the series.
+	// when a write is made, the total number of points in the series
+	// will be sent over the channel, so we can make sure we are reading
+	// the correct amount of values.
+	go func(numPoints chan int, finished chan struct{}) {
+		n := 0
+		for {
+			select {
+			case n = <-numPoints:
+				//new point has been written. total points in now "n"
+			default:
```

Comment:
This is deliberate. If 'go test -race' doesn't complain then it is a certainty that there are no race conditions.  Additionally high CPU is desired for this test to catch situations where writes happen while the goroutine is reading data.